### PR TITLE
ADCM-6118: Implement Host object

### PR DESCRIPTION
**Pull request by Starovoitov**

### CODE REVIEW

File: adcm_aio_client/core/objects/cm.py [hunk 1]
```diff
@@ -110,3 +116,56 @@ def get_own_path(self: Self) -> Endpoint:
...
+    def name(self: Self) -> str:
+        return str(self._data["name"])
+
+    @property
+    def description(self: Self) -> str:
+        return str(self._data["description"])
+
+    @property
+    def status(self: Self) -> ADCMEntityStatus:
+        return ADCMEntityStatus(self._data["status"])
```
Comment (a-alferov) [inline adcm_aio_client/core/objects/cm.py:142]: Change pls this method on `get_status`. See example in `Cluster` object

File: adcm_aio_client/core/objects/cm.py [hunk 2]
```diff
@@ -110,3 +116,56 @@ def get_own_path(self: Self) -> Endpoint:
...
     class_type = Component
+
+
+class ConfigsNode(PaginatedChildAccessor): ...
+
+
+class HostProvidersNode(PaginatedChildAccessor): ...
+
+
+class Host(InteractiveChildObject[Cluster | None]):
```
Comment (Sealwing) [inline adcm_aio_client/core/objects/cm.py:127]: `Cluster` is not a parent of `Host` (I see that you've tried to cover case with "Host of Cluster", but IMO it's too early for such stretch)

File: adcm_aio_client/core/objects/cm.py [hunk 3]
```diff
@@ -110,3 +116,56 @@ def get_own_path(self: Self) -> Endpoint:
...
+    @property
+    def status(self: Self) -> ADCMEntityStatus:
+        return ADCMEntityStatus(self._data["status"])
+
+    async def get_status(self: Self) -> ADCMEntityStatus:
+        response = await self._requester.get(*self.get_own_path())
+        return ADCMEntityStatus(response.as_dict()["status"])
+
+    @cached_property
+    def cluster(self: Self) -> ClustersNode | None:
```
Comment (Sealwing) [inline adcm_aio_client/core/objects/cm.py:149]: Not a `Node`, `Cluster`

**Review by Sealwing — CHANGES_REQUESTED** [review]

File: adcm_aio_client/core/objects/cm.py (hunk 1 above)
Comment (Starovoitov) [reply adcm_aio_client/core/objects/cm.py]: changed

File: adcm_aio_client/core/objects/cm.py (hunk 2 above)
Comment (Starovoitov) [reply adcm_aio_client/core/objects/cm.py]: changed

File: adcm_aio_client/core/objects/cm.py (hunk 3 above)
Comment (Starovoitov) [reply adcm_aio_client/core/objects/cm.py]: changed

File: adcm_aio_client/core/objects/cm.py [hunk 4]
```diff
@@ -117,13 +113,51 @@ class ServicesNode(PaginatedChildAccessor[Cluster, Service, None]):
...
+        return ADCMEntityStatus(response.as_dict()["status"])
+
+    @cached_property
+    def cluster(self: Self) -> Cluster | None:
+        if not self._data["cluster"]:
+            return None
+        return Cluster(requester=self._requester, data=self._data["cluster"])
+
+    @cached_property
+    def hostprovider(self: Self) -> "HostProvidersNode":
```
Comment (Sealwing) [inline adcm_aio_client/core/objects/cm.py:149]: not a node, object

File: adcm_aio_client/core/objects/cm.py [hunk 5]
```diff
@@ -117,13 +113,51 @@ class ServicesNode(PaginatedChildAccessor[Cluster, Service, None]):
...
 
 
 class ComponentsNode(NonPaginatedChildAccessor[Service, Component, None]):
     class_type = Component
+
+
+class HostProvidersNode(PaginatedChildAccessor): ...
+
+
+class Host(Deletable, InteractiveChildObject):
```
Comment (Sealwing) [inline adcm_aio_client/core/objects/cm.py:128]: `Host` is not a child of anything, it's self-sufficient: Accessible from client's root

**Review by Sealwing — CHANGES_REQUESTED** [review]

File: adcm_aio_client/core/objects/cm.py [hunk 6]
```diff
@@ -117,13 +113,51 @@ class ServicesNode(PaginatedChildAccessor[Cluster, Service, None]):
...
+
+    async def get_status(self: Self) -> ADCMEntityStatus:
+        response = await self._requester.get(*self.get_own_path())
+        return ADCMEntityStatus(response.as_dict()["status"])
+
+    @cached_property
+    def cluster(self: Self) -> Cluster | None:
+        if not self._data["cluster"]:
+            return None
+        return Cluster(requester=self._requester, data=self._data["cluster"])
```
Comment (Sealwing) [inline adcm_aio_client/core/objects/cm.py:146]: `cluster` info is not enough here, should be re-requested by cluster id, same for hostprovider

Comment (Sealwing) [reply adcm_aio_client/core/objects/cm.py:146]: Let's expand `RootInteractiveObject`:

```python
# we have this
class RootInteractiveObject(InteractiveObject):
    PATH_PREFIX: str

    def get_own_path(self: Self) -> Endpoint:
        # change here
        return self._build_own_path(self.id)

    # let's add this one
    @classmethod
    async def with_id(cls: type[Self], requester: Requester, id: int) -> Self:
        object_path = cls._build_own_path(id)
        response = await requester.get(*object_path)
        return cls(requester=requester, data=response.as_dict()) 

    @classmethod
    def _build_own_path(cls, id: int) -> Endpoint:
        return cls.PATH_PREFIX, id

# and now we can use it in methods like cluster

class Host:
    ...
    @cached_property
    async def cluster(self) -> Cluster | None:
        if notindata:
            return None

        return await Cluster.with_id(requester=self._requester, id=self._data["cluster"]["id"]
```

File: adcm_aio_client/core/objects/cm.py (hunk 4 above)
Comment (Starovoitov) [reply adcm_aio_client/core/objects/cm.py]: fixed

File: adcm_aio_client/core/objects/cm.py (hunk 5 above)
Comment (Starovoitov) [reply adcm_aio_client/core/objects/cm.py]: fixed

File: adcm_aio_client/core/objects/cm.py [hunk 7]
```diff
@@ -117,13 +113,51 @@ class ServicesNode(PaginatedChildAccessor[Cluster, Service, None]):
...
+
+    @cached_property
+    def hostprovider(self: Self) -> "HostProvidersNode":
+        return HostProvidersNode(parent=None, path=(*self.get_own_path(), "hostprovider"), requester=self._requester)
+
+    def get_own_path(self: Self) -> Endpoint:
+        return "hosts", self.id
+
+
+class HostsNode(PaginatedAccessor[Host, None]):
```
[the remark this answers is not in the document]
Comment (Starovoitov) [reply adcm_aio_client/core/objects/cm.py:160]: added

File: adcm_aio_client/core/objects/cm.py (hunk 6 above)
Comment (Starovoitov) [reply adcm_aio_client/core/objects/cm.py]: done

**Review by Sealwing — APPROVED** [review]

**Review by a-alferov — APPROVED** [review]

**Pull request merged**